### PR TITLE
added ./bin/sessionmirror (adapted from xnatmirror)

### PR DESCRIPTION
**Pull request by xgrg**

### CODE REVIEW

File: bin/sessionmirror.py [hunk 1]
```diff
@@ -0,0 +1,749 @@
...
+    '''Check if resource contains any files'''
+    f_count = 0
+    for f_in in _res.files().fetchall('obj'):
+        f_count += 1
+        break
+
+    return f_count == 0
+
+
+def copy_project(src_proj, dst_proj, proj_cache_dir):
```
Comment (jhuguetn) [inline bin/sessionmirror.py]: Note that `copy_project` function does not copy project-level resources.

File: bin/sessionmirror.py [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+        subject_label = subj['label']
+
+        print("INFO:Processing subject %s (%d)..." % (subject_label, subj_i))
+        src_subj = src_proj.subject(subject_label)
+        dst_subj = dst_proj.subject(subject_label)
+        subj_cache_dir = op.join(proj_cache_dir, subject_label)
+        copy_subject(src_subj, dst_subj, subj_cache_dir)
+
+
+def copy_subject(src_subj, dst_subj, subj_cache_dir):
```
Comment (jhuguetn) [inline bin/sessionmirror.py]: Note that `copy_subject` function does not copy subject-level resources

File: bin/sessionmirror.py [hunk 3]
```diff
@@ -0,0 +1,749 @@
...
+    if is_empty:
+        if use_zip:
+            # Try to copy as zip
+            try:
+                print('INFO:Copying resource as zip: %s...' % src_res.label())
+                copy_res_zip(src_res, dst_res, res_cache_dir)
+                return
+            except Exception:
+                try:
+                    print('INFO: second attempt to copy resource as zip: %s...'
```
Comment (jhuguetn) [inline bin/sessionmirror.py]: Why there are 2 attempts to upload the bundled contents of `resource` using zipped files but solely one when following a file-by-file approach?

File: bin/sessionmirror.py [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+        ET.register_namespace('xnat', 'http://nrg.wustl.edu/xnat')
+        ET.register_namespace('proc', 'http://nrg.wustl.edu/proc')
+        ET.register_namespace('prov', 'http://www.nbirn.net/prov')
+        ET.register_namespace('fs', 'http://nrg.wustl.edu/fs')
+        ET.ElementTree(root).write(file_path)
+    except IOError as error:
+        print('ERROR:writing xml file: {}: {}'.format(file_path, str(error)))
+
+
+def check_attributes(src_obj, dest_obj, dtype=None):
```
Comment (jhuguetn) [inline bin/sessionmirror.py]: function `check_attributes` is defined but seems to never be used

File: bin/sessionmirror.py [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+
+def main(args):
+    x1 = pyxnat.Interface(config=args.source_config)
+    x2 = pyxnat.Interface(config=args.dest_config)
+
+    columns = ['subject_label', 'label']
+    e1 = x1.array.experiments(experiment_id=args.experiment_id,
+        columns=columns).data[0]
+    p = x2.select.project(args.project_id)
+    s = p.subject(e1['subject_label'])
```
Comment (jhuguetn) [inline bin/sessionmirror.py]: what if subject exists?

File: requirements.txt [hunk 6]
```diff
@@ -1,2 +1,3 @@
 lxml>=4.3.2
 requests>=2.20
+future>=0.16
```
Comment (jhuguetn) [inline requirements.txt]: where is that new requirement being used?

File: bin/sessionmirror.py (hunk 1 above)
Comment (xgrg) [reply bin/sessionmirror.py]: Right. Actually this function was just copied/pasted. This is not tested nor used by `sessionmirror.py`, as this only copies sessions (the original `xnat_mirror.py` is intended for entire projects). But you are right, project-level resources seem to be ignored. We may also just remove them for the moment.

File: bin/sessionmirror.py (hunk 2 above)
Comment (xgrg) [reply bin/sessionmirror.py]: Same comment as for `copy_project`.

File: bin/sessionmirror.py (hunk 3 above)
Comment (xgrg) [reply bin/sessionmirror.py]: I would say for the same reason as we have introduced `STUBBORN` in other parts of `pyxnat`.

File: bin/sessionmirror.py (hunk 4 above)
Comment (xgrg) [reply bin/sessionmirror.py]: This part was hard to follow in the original code (mostly commented out).
(see https://github.com/VUIIS/dax/blob/master/bin/Xnat_tools/Xnatmirror#L461)

Again we may just remove the function if not used.

File: requirements.txt (hunk 6 above)
Comment (xgrg) [reply requirements.txt]: At the beginning: 

```
from builtins import zip
from builtins import input
from builtins import str
```

File: bin/sessionmirror.py (hunk 5 above)
Comment (xgrg) [reply bin/sessionmirror.py]: > as long as experiment does not exist
Experiment never exists since it is created/duplicated.

> what if subject exists?
After a thorough analysis of the code (sic), it appears that if subject exists, nothing happens.

File: requirements.txt (hunk 6 above)
Comment (jhuguetn) [reply requirements.txt]: I see, thanks for clarifying

File: bin/sessionmirror.py (hunk 5 above)
Comment (jhuguetn) [reply bin/sessionmirror.py]:
```suggestion
    s = p.subject(e1['subject_label'])
    if not s.exists():
        s.create()
```

File: bin/sessionmirror.py (hunk 3 above)
Comment (jhuguetn) [reply bin/sessionmirror.py]: odd but ok

File: bin/sessionmirror.py (hunk 4 above)
Comment (jhuguetn) [reply bin/sessionmirror.py]: I see, is a sort of additional control mechanism for checking the consistency of the resource attributes. We could exclude it easing the maintenance of ported code but keeping the function as is now is not hurting anyway.

**Comment by jhuguetn** [comment]
Check my last comments, overall is all good to merge your PR in my opinion. Good job @xgrg!
